### PR TITLE
Compute mutability of type before processing specifications

### PR DESCRIPTION
**Pull request by n-osborne**
Traverse a type definition to compute whether it is an OCaml mutable type. 
This allows to have information even in the absence of specifications.

I also add some tests.

### CODE REVIEW

**Comment by pascutto** [comment]
I'm a bit torn about the mutability status of `string`. I'm tempted to consider it as mutable, since the `String` module exposes the `unsafe_to_bytes` function, which lets you modify it. On the other hand this function is explicitly `unsafe`, so I'm not sure what we should do with this.
Do you have an opinion on this?

**Comment by n-osborne** [comment]
The documentation says that `string` is not mutable anymore, except with a special option given to the compiler for compatibility purposes.
As it is still possible to have mutable strings, if we say they are not, we are making an affirmation about good practice. Per haps it is no the place of `ortac` to do so?

**Comment by pascutto** [comment]
`string`s themselves are not mutable, but they can be converted to `bytes`, which are mutable. When using the `unsafe` convertion functions to do so, the underlying bytes are not copied, which means you can modify the `string` through the converted `byte`. e.g.
```ocaml
let s = "abc" in
let b = Bytes.unsafe_of_string s in
Bytes.set b 0 '0';
print_endline s
```
is valid without any compiler option, and prints `"0bc"`.

I guess the same kind of broken mutability can also happen with `Obj.magic` with basically any type. So maybe `ortac` can indeed ignore that, and assume that the program does not use unsafe functions (we'll have to write that somewhere in the documentation).

File: src/core/translate.ml [hunk 1]
```diff
@@ -53,11 +53,15 @@ let var_of_arg ~driver arg : Translated.ocaml_var =
...
   let name = td.td_ts.ts_ident.id_str in
   let loc = td.td_loc in
-  let mutable_ = false (* XXX FIXME look at the ocaml type *) in
+  let mutable_ = Mutability.type_declaration ~driver td in
   let type_ = type_ ~name ~loc ~mutable_ ~ghost in
   let process ~type_ (spec : Tast.type_spec) =
     let term_printer = Fmt.str "%a" Tterm.print_term in
-    let mutable_ = type_.mutable_ || spec.ty_ephemeral in
+    let mutable_ =
+      max
```
Comment (pascutto) [inline src/core/translate.ml:61]: I think using the polymorphic `max` is a bit fragile here. Perhaps you can define an explicit one in `Mutability`?

File: src/core/types.ml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    match t.ty_node with
+    | Tyvar _ -> Translated.Unknown
+    | Tyapp (ts, tyl) ->
+        List.fold_left
+          (fun acc t -> max acc (ty ~driver t))
+          (tysymbol ~driver ts) tyl
+
+  let lsymbol ~driver (ls : Tterm.lsymbol) =
+    Option.fold ~none:Translated.Unknown ~some:(ty ~driver) ls.ls_value
+    |> List.fold_right (fun t acc -> max acc (ty ~driver t)) ls.ls_args
```
Comment (pascutto) [inline src/core/types.ml:28]: I think you could `fold_left` instead to get the tail-recursiveness (not that it matter that much with these small lists...).

**Review by pascutto — COMMENTED** [review]
I think the `max` part is critical, apart from that it looks great!

File: src/core/types.ml (hunk 2 above)
Comment (n-osborne) [reply src/core/types.ml:28]: It was for the pipe.

**Pull request merged**